gateway/ocfl_repository_gateway.py

```diff
@@ -0,0 +1,128 @@
+    ObjectAlreadyExistsError,
+    ObjectDoesNotExistError,
+    ObjectNotStagedError,
+    RepositoryGatewayError
+)
+from gateway.object_file import ObjectFile
+from gateway.package import Package
+from gateway.repository_gateway import RepositoryGateway
+
+logger = logging.getLogger(__file__)
+
+class OcflRepositoryGateway(RepositoryGateway):
+    storage_layout = "0002-flat-direct-storage-layout"
+
+    def __init__(self, storage_path: Path):
+        self.storage_path: Path = storage_path
+
+    def create_repository(self) -> None:
+        args = [
+            "rocfl", "-r", self.storage_path, "init",
+            "-l", self.storage_layout
+        ]
+        try:
+            subprocess.run(args, check=True, capture_output=True)
+        except CalledProcessError as e:
+            raise RepositoryGatewayError() from e
+
+    def create_staged_object(self, id: str) -> None:
+        args = ["rocfl", "-r", self.storage_path, "new", id]
+        try:
+            subprocess.run(args, check=True, capture_output=True)
+        except CalledProcessError as e:
+            already_exists_message = (
+                f"Illegal state: Cannot create object {id} because it already exists in staging"
+            )
+            if already_exists_message in e.stderr.decode():
+                raise ObjectAlreadyExistsError()
+            raise RepositoryGatewayError() from e
+
+    def stage_object_files(self, id: str, source_package: Package) -> None:
+        command = " ".join([
+            "rocfl", "-r", str(self.storage_path),
+            "cp", "-r", id, str(source_package.get_root_path() / "*"), "--", "/"
+        ])
+        try:
+            subprocess.run(command, check=True, shell=True, capture_output=True)
+        except CalledProcessError as e:
+            not_found_message = f"Not found: Object {id}"
+            if not_found_message in e.stderr.decode():
+                raise ObjectDoesNotExistError()
+            raise RepositoryGatewayError() from e
+
+    def commit_object_changes(
+        self,
+        id: str,
+        coordinator: Coordinator,
+        message: str
+    ) -> None:
+        args = [
+            "rocfl", "-r", self.storage_path, "commit", id,
+            "-n", coordinator.username, "-a", f"mailto:{coordinator.email}",
+            "-m", message
+        ]
+        try:
+            subprocess.run(args, check=True, capture_output=True)
+        except CalledProcessError as e:
+            no_staged_changes_message = f"No staged changes found for object {id}"
+            if no_staged_changes_message in e.stderr.decode():
+                raise ObjectNotStagedError()
+            raise RepositoryGatewayError() from e
+
+    def purge_object(self, id: str) -> None:
+        args = ["rocfl", "-r", self.storage_path, "purge", "-f", id]
+        try:
+            subprocess.run(args, check=True, capture_output=True)
+        except CalledProcessError as e:
+            raise RepositoryGatewayError() from e
```

Comment:
I found some weird behavior when testing; I don't think this is raises an error if an object isn't found. Do we want raise any errors ourselves if a user tries to do this, or ...?

Reply:
Probably not since it could be a repeated call to purge … CRUD has a similar behavior for Delete.

Reply:
Database deletes do not raise exceptions; shold this behave like `rm -r` or `rm -rf`? I think the latter.

Reply:
Ah! That makes sense. Seems like `rm -rf` also avoids error codes, so perhaps this is just designed to mimic that behavior. I think without the `-f` here, it prompts, which I don't think we have a way to handle easily.